ScrollTo API on List widget

AnkushJadhav opened the pull request:
<!-- If this is your first pull request for Fyne please read the contributor docs at:
https://github.com/fyne-io/fyne/wiki/Contributing.
Be sure that your work is based off `develop` branch. --> 

### Description:
Exposed a `ScrollTo` API that scrolls a List widget and sets the focus on the ListItem represented by the id given to the API.

Fixes #2105 

### Checklist:
<!-- Please tick these as appropriate using [x] -->

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

- [x] Public APIs match existing style.

Jacalz commented on widget/list.go:
```diff
@@ -95,6 +95,19 @@ func (l *List) MinSize() fyne.Size {
+	if l.scroller == nil {
+		return
+	}
+	y := (float32(id) * l.itemMin.Height) + (float32(id) * theme.SeparatorThicknessSize())
```
I know that it was like this before, but this can be simplified. You can rewrite it as the following:
```suggestion
	y := float32(id) * (l.itemMin.Height) + theme.SeparatorThicknessSize())
```

Jacalz replied:
Out of interest, I double checked to see if the resulting assembly were different or if the compiler would be smart enough to optimize it away. It does look like this change indeed does save us a multiplication, as I expected 🙂 https://go.godbolt.org/z/1556hfGnr

AnkushJadhav replied:
That's a great catch! Thanks for the suggestion!

Jacalz replied:
You're welcome. Feels a bit strange though that the compiler wouldn't be able to optimize it away. Opened an issue over at golang/go, https://github.com/golang/go/issues/45760.

Jacalz replied:
Comments in that issue mentioned that rounding will make this create different values. https://play.golang.org/p/ziUoFO0OhNT

I think you might want to review this change. Sorry 😬

Jacalz replied:
Oh, sorry looks like there was a typo in that message. I meant that you might need to revert the change. It can cause issues with rounding and be less precise that what it was before.

Jacalz replied:
Thanks for reverting it back to how it was. You can remove the parenthesis if you like, but now exactly needed.

AnkushJadhav replied:
I prefer parenthesis from a code readability perspective. Can remove them if it goes against the project design guidelines.

Jacalz replied:
I don’t think it goes against any guidelines, no. I just feel that they are unnecessary given that order of operation means that they have no meaning. This is not really hard to read code in the first place, but if you want to keep them, feel free to do so. I’m not going to stop you 🙂

Jacalz reviewed (changes requested):
Nick work. Just a little suggestion to simplify a math operation.

Jacalz reviewed (dismissed):
Thanks. This looks good to me 👍 

Just one little note. This seems to incorrectly mention fixing https://github.com/fyne-io/fyne/issues/1892 which is specific to the table widget. Did you plan to add in a fix for that as well?

andydotxyz reviewed (changes requested):
An api “ScrollTo” should not change selection, there is a separate API that does that already. 

Also the initial issue requests ScrollToBottom, would that and ScrollToTop be useful too?

AnkushJadhav commented:
> Thanks. This looks good to me 👍
> 
> Just one little note. This seems to incorrectly mention fixing #1892 which is specific to the table widget. Did you plan to add in a fix for that as well?

@Jacalz No, I think I added that one incorrectly. Not sure what I was thinking. Updated the PR description.

AnkushJadhav commented:
> An api “ScrollTo” should not change selection, there is a separate API that does that already.
> 
> Also the initial issue requests ScrollToBottom, would that and ScrollToTop be useful too?

@andydotxyz The "ScrollTo" API does not change the selection. It simply scrolls to the item provided to it. Any selection made on the list remains intact.
A ScrollToTop and ScrollToBottom seemed more like utility functions to me, that don't add much value. The developer can easily implement these functions themselves by using the ScrollTo API. Your thoughts?

AnkushJadhav commented:
@andydotxyz The existing `Select` API modified the selection and also scrolled to the selection. While the new `ScrollTo` API ONLY scrolls to the item. Since both these APIs have a "scroll to item" kind of logic in them, so extracted that logic into a common helper function.

andydotxyz commented on widget/list.go:
```diff
@@ -95,6 +95,19 @@ func (l *List) MinSize() fyne.Size {
 	return l.BaseWidget.MinSize()
 }
 
+func (l *List) moveFocusTo(id ListItemID) {
```
This name confused me. Please don't use the word "Focus" when you are referring to visibility

andydotxyz reviewed (changes requested):
I understand now - apologies that I misread the code.
I was confused by the use of the word "Focus" when you were just meaning scroll to make visible. Can you please update the method name so it does not imply focus change.

andydotxyz commented:
> @andydotxyz The existing `Select` API modified the selection and also scrolled to the selection. While the new `ScrollTo` API ONLY scrolls to the item. Since both these APIs have a "scroll to item" kind of logic in them, so extracted that logic into a common helper function.

You're quite right, apologies. The method name tripped me up, can you find a better name?

andydotxyz commented:
> A ScrollToTop and ScrollToBottom seemed more like utility functions to me, that don't add much value

I was just suggesting it because those were added to `container.Scroll` at the request of previous users.

AnkushJadhav commented:
> > A ScrollToTop and ScrollToBottom seemed more like utility functions to me, that don't add much value
> 
> I was just suggesting it because those were added to `container.Scroll` at the request of previous users.

Oh, then it makes sense to have them for consistency. Updated the PR.

andydotxyz commented on widget/list.go:
```diff
@@ -147,6 +147,22 @@ func (l *List) ScrollTo(id ListItemID) {
+	if f := l.Length; f != nil {
+		length = f()
+	}
+	l.scrollTo(length)
```
should this not be `length - 1` (as long as length is > 0)?

andydotxyz reviewed (changes requested):
I think there is an off-by-one in there.

andydotxyz reviewed (approved):
Great, thanks

andydotxyz commented:
It seems that GitHub has a policy or blocking first-time contributor CI. I have asked it to progress now.
As a result there are some staticcheck errors to fix, sorry!

andydotxyz reviewed (changes requested):
I am so sorry for forgetting about this (my head was in bug fixing) but each of the new methods needs the following comment added:
```go
//
// Since: 2.1
```

AnkushJadhav commented:
@andydotxyz Thanks for pointing that out. Sorry, this is my first time on a big open source project, so still learning all the nitty-gritty.

andydotxyz commented:
> @andydotxyz Thanks for pointing that out. Sorry, this is my first time on a big open source project, so still learning all the nitty-gritty.

Don't apologise, we have a lot of processes to keep the API need and consistent. There is a load written down at https://github.com/fyne-io/fyne/wiki/Contributing
